File: examples/csharp/SolidJobsClient.cs
using System.Globalization;
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using SolidJobs.Models;

namespace SolidJobs;

/// <summary>
/// Reusable client for the SolidJobs public offers API. Wraps <see cref="HttpClient"/>:
/// builds the query string, deserializes into strongly typed models and handles
/// 429 (Retry-After + exponential backoff).
/// </summary>
public sealed partial class SolidJobsClient
{
    private static readonly Regex CampaignRegex = CampaignRegexCompiled();

    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        Converters = { new JsonStringEnumConverter() },
    };

    private readonly HttpClient http;
    private readonly int maxRetries;

    /// <param name="http">Shared HttpClient (from IHttpClientFactory).</param>
    /// <param name="maxRetries">Number of retries when the API responds with 429.</param>
    public SolidJobsClient(HttpClient http, int maxRetries = 3)
    {
        this.http = http;
        this.maxRetries = maxRetries;
    }

    /// <summary>Fetches a single page of offers for the given division.</summary>
    public async Task<OffersResponse> GetOffersAsync(
        Division division,
        string campaign,
        OfferQuery? query = null,
        CancellationToken cancellationToken = default)
    {
        if (!CampaignRegex.IsMatch(campaign))
        {
            throw new ArgumentException(
                "Campaign must contain only lowercase letters, digits and hyphens (max 64 chars).",
                nameof(campaign));
        }

        var requestUri = BuildRequestUri(division, campaign, query ?? new OfferQuery());

        using var response = await SendWithRetryAsync(requestUri, cancellationToken);
        response.EnsureSuccessStatusCode();

        var result = await response.Content.ReadFromJsonAsync<OffersResponse>(JsonOptions, cancellationToken);

        return result ?? throw new InvalidOperationException("API returned an empty response body.");
    }

    /// <summary>Fetches labour-market statistics for the given scope.</summary>
    /// <param name="scopeKind">Scope kind: <c>division</c>, <c>mainCategory</c>, <c>subcategory</c>, <c>subcategoryGroup</c> or <c>city</c>.</param>
    /// <param name="scopeKey">Scope key within the kind (e.g. <c>IT</c>, <c>React</c>, <c>warszawa</c>).</param>
    /// <param name="campaign">Traffic-tracking identifier (required).</param>
    /// <param name="fields">Optional subset of sections to return; null/empty returns all sections available for the scope.</param>
    public async Task<MarketStatisticsResponse> GetMarketStatisticsAsync(
        string scopeKind,
        string scopeKey,
        string campaign,
        IReadOnlyList<string>? fields = null,
        CancellationToken cancellationToken = default)
    {
        if (!CampaignRegex.IsMatch(campaign))
        {
            throw new ArgumentException(
                "Campaign must contain only lowercase letters, digits and hyphens (max 64 chars).",
                nameof(campaign));
        }

        var query = $"campaign={Uri.EscapeDataString(campaign)}";
        if (fields is { Count: > 0 })
        {
            query += $"&fields={Uri.EscapeDataString(string.Join(',', fields))}";
        }

        var requestUri = $"public-api/market-statistics/{Uri.EscapeDataString(scopeKind)}/{Uri.EscapeDataString(scopeKey)}?{query}";

        using var response = await SendWithRetryAsync(requestUri, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            var body = await response.Content.ReadAsStringAsync(cancellationToken);
            throw new HttpRequestException($"HTTP {(int)response.StatusCode}: {body}");
        }

        var result = await response.Content.ReadFromJsonAsync<MarketStatisticsResponse>(JsonOptions, cancellationToken);

        return result ?? throw new InvalidOperationException("API returned an empty response body.");
    }

    /// <summary>Fetches the yearly market report for a single role.</summary>
    /// <param name="scopeKey">Role (specialization) the report describes, e.g. <c>ManualTester</c>.</param>
    /// <param name="campaign">Traffic-tracking identifier (required).</param>
    /// <remarks>
    /// Unlike <see cref="GetMarketStatisticsAsync"/> there is no scope kind and no <c>fields</c> filter —
    /// the report always comes back whole, spanning up to 3 calendar years, oldest first.
    /// </remarks>
    public async Task<MarketRaportResponse> GetMarketRaportAsync(
        string scopeKey,
        string campaign,
        CancellationToken cancellationToken = default)
    {
        if (!CampaignRegex.IsMatch(campaign))
        {
            throw new ArgumentException(
                "Campaign must contain only lowercase letters, digits and hyphens (max 64 chars).",
                nameof(campaign));
        }

        var query = $"campaign={Uri.EscapeDataString(campaign)}";
        var requestUri = $"public-api/market-statistics/raport/{Uri.EscapeDataString(scopeKey)}?{query}";

        using var response = await SendWithRetryAsync(requestUri, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            var body = await response.Content.ReadAsStringAsync(cancellationToken);
            throw new HttpRequestException($"HTTP {(int)response.StatusCode}: {body}");
        }

        var result = await response.Content.ReadFromJsonAsync<MarketRaportResponse>(JsonOptions, cancellationToken);

        return result ?? throw new InvalidOperationException("API returned an empty response body.");
    }

    /// <summary>
    /// Streams every offer matching the criteria, walking the pages automatically.
    /// </summary>
    public async IAsyncEnumerable<JobOffer> GetAllOffersAsync(
        Division division,
        string campaign,
        OfferQuery? query = null,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        var baseQuery = query ?? new OfferQuery();
        var pageIndex = baseQuery.PageIndex;

        while (true)
        {
            var page = await GetOffersAsync(
                division,
                campaign,
                baseQuery with { PageIndex = pageIndex },
                cancellationToken);

            foreach (var job in page.Jobs)
            {
                yield return job;
            }

            pageIndex++;

            if (page.Jobs.Count == 0 || pageIndex >= page.TotalPages)
            {
                yield break;
            }
        }
    }

    private async Task<HttpResponseMessage> SendWithRetryAsync(string requestUri, CancellationToken cancellationToken)
    {
        for (var attempt = 0; ; attempt++)
        {
            var response = await http.GetAsync(requestUri, cancellationToken);

            if (response.StatusCode is not HttpStatusCode.TooManyRequests || attempt >= maxRetries)
            {
                return response;
            }

            var delay = GetRetryDelay(response.Headers.RetryAfter, attempt);
            response.Dispose();

            await Task.Delay(delay, cancellationToken);
        }
    }

    private static TimeSpan GetRetryDelay(RetryConditionHeaderValue? retryAfter, int attempt)
    {
        if (retryAfter?.Delta is { } delta)
        {
            return delta;
        }

        if (retryAfter?.Date is { } date)
        {
            var until = date - DateTimeOffset.UtcNow;
            if (until > TimeSpan.Zero)
            {
                return until;
            }
        }

        return TimeSpan.FromSeconds(Math.Pow(2, attempt));
    }

    private static string BuildRequestUri(Division division, string campaign, OfferQuery query)
    {
        var parameters = new List<KeyValuePair<string, string>>
        {
            new("campaign", campaign),
        };

        if (query.PageIndex > 0)
        {
            parameters.Add(new("pageIndex", query.PageIndex.ToString(CultureInfo.InvariantCulture)));
        }

        if (query.PageSize is { } pageSize)
        {
            parameters.Add(new("pageSize", pageSize.ToString(CultureInfo.InvariantCulture)));
        }

        if (query.Sort is { } sort)
        {
            parameters.Add(new("sortActive", ToSortToken(sort)));
            parameters.Add(new("sortDirection", query.Direction is SortDirection.Desc ? "desc" : "asc"));
        }

        if (query.Cities.Count > 0)
        {
            parameters.Add(new("search.cities", string.Join(',', query.Cities)));
        }

        AddRepeated(parameters, "search.categories", query.Categories.Select(c => c.ToString()));
        AddRepeated(parameters, "search.subCategories", query.SubCategories.Select(sc => sc.ToString()));
        AddRepeated(parameters, "search.experiences", query.Experiences.Select(e => e.ToString()));
        AddRepeated(parameters, "search.searchTerm", query.SearchTerms);

        if (query.MinimumSalary is { } minimumSalary)
        {
            parameters.Add(new("search.minimumSalary", minimumSalary.ToString(CultureInfo.InvariantCulture)));
        }

        var queryString = string.Join(
            '&',
            parameters.Select(p => $"{Uri.EscapeDataString(p.Key)}={Uri.EscapeDataString(p.Value)}"));

        return $"public-api/offers/{division}?{queryString}";

        static void AddRepeated(List<KeyValuePair<string, string>> target, string key, IEnumerable<string> values)
        {
            foreach (var value in values)
            {
                target.Add(new(key, value));
            }
        }
    }

    private static string ToSortToken(SortField sort) => sort switch
    {
        SortField.ValidFrom => "validFrom",
        SortField.ValidTo => "validTo",
        SortField.Title => "title",
        SortField.Company => "company",
        SortField.SalaryFrom => "salaryFrom",
        SortField.SalaryTo => "salaryTo",
        SortField.ExperienceLevel => "experienceLevel",
        _ => "validFrom",
    };

    [GeneratedRegex(@"^[a-z0-9-]{1,64}$", RegexOptions.Compiled)]
    private static partial Regex CampaignRegexCompiled();
}

File: examples/java/SolidJobsClient.java
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.StringJoiner;
import java.util.regex.Pattern;

/**
 * Reusable client for the SolidJobs public API.
 * Uses built-in java.net.http (Java 11+), no external dependencies.
 */
public class SolidJobsClient {

    static final String BASE_URL = "https://solid.jobs";
    static final String API_VERSION = "1.0";
    static final int MAX_RETRIES = 3;
    static final Pattern CAMPAIGN_RE = Pattern.compile("^[a-z0-9-]{1,64}$");

    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    public String fetchOffers(String division, String campaign, Map<String, String> extraParams) throws Exception {
        if (!CAMPAIGN_RE.matcher(campaign).matches()) {
            throw new IllegalArgumentException("Invalid campaign format.");
        }

        var uri = buildUri(division, campaign, extraParams);

        for (int attempt = 0; ; attempt++) {
            var request = HttpRequest.newBuilder()
                    .uri(URI.create(uri))
                    .header("Accept", "application/json")
                    .header("X-Api-Version", API_VERSION)
                    .timeout(Duration.ofSeconds(30))
                    .build();

            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 429 || attempt >= MAX_RETRIES) {
                if (response.statusCode() != 200) {
                    throw new RuntimeException("HTTP " + response.statusCode() + ": " + response.body());
                }
                return response.body();
            }

            var retryAfter = response.headers().firstValue("Retry-After");
            long delay = retryAfter.map(Long::parseLong).orElse((long) Math.pow(2, attempt));
            System.err.printf("Rate limited (429). Retrying in %ds...%n", delay);
            Thread.sleep(delay * 1000);
        }
    }

    public String fetchMarketStatistics(String scopeKind, String scopeKey, String campaign, String fields) throws Exception {
        if (!CAMPAIGN_RE.matcher(campaign).matches()) {
            throw new IllegalArgumentException("Invalid campaign format.");
        }

        var query = new StringJoiner("&");
        query.add("campaign=" + encode(campaign));
        if (fields != null && !fields.isEmpty()) {
            query.add("fields=" + encode(fields));
        }
        var uri = BASE_URL + "/public-api/market-statistics/" + encode(scopeKind) + "/" + encode(scopeKey) + "?" + query;

        for (int attempt = 0; ; attempt++) {
            var request = HttpRequest.newBuilder()
                    .uri(URI.create(uri))
                    .header("Accept", "application/json")
                    .header("X-Api-Version", API_VERSION)
                    .timeout(Duration.ofSeconds(30))
                    .build();

            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 429 || attempt >= MAX_RETRIES) {
                if (response.statusCode() != 200) {
                    throw new RuntimeException("HTTP " + response.statusCode() + ": " + response.body());
                }
                return response.body();
            }

            var retryAfter = response.headers().firstValue("Retry-After");
            long delay = retryAfter.map(Long::parseLong).orElse((long) Math.pow(2, attempt));
            System.err.printf("Rate limited (429). Retrying in %ds...%n", delay);
            Thread.sleep(delay * 1000);
        }
    }

    private static String buildUri(String division, String campaign, Map<String, String> params) {
        var sb = new StringJoiner("&");
        sb.add("campaign=" + encode(campaign));
        for (var entry : params.entrySet()) {
            sb.add(encode(entry.getKey()) + "=" + encode(entry.getValue()));
        }
        return BASE_URL + "/public-api/offers/" + division + "?" + sb;
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }
}
